### PR TITLE
Start Full Scan from a random index for Least Request LB.

### PR DESCRIPTION
**Pull request by barroca**
Fixed a bug (https://github.com/envoyproxy/envoy/pull/11006) that caused the Least Request load balancer policy to choose the first host of the list when the number of requests are the same during a full scan. Start the selection from a random index instead of 0.

### CODE REVIEW

**Review by tomwans — APPROVED** [review]

**Review by tonya11en — CHANGES_REQUESTED** [review]
@barroca can you elaborate on how starting from a random index fixes the problem in https://github.com/envoyproxy/envoy/issues/11004 and what  https://github.com/envoyproxy/envoy/pull/11006 tried to address? What effect is this change trying to have on the selection probabilities?

We've had many discussions on this (see https://github.com/envoyproxy/envoy/pull/11006#issuecomment-627548703) that lead to us leaving the P2C algorithm alone. Attempting to force a full scan, while guaranteeing selection of the host with the lowest active requests, opens the system up to [herding behavior](https://en.wikipedia.org/wiki/Thundering_herd_problem). While there are [some applications](https://github.com/envoyproxy/envoy/issues/4481) where this would be useful, one of the key insights of the [Azar et. al. paper](https://homes.cs.washington.edu/~karlin/papers/AzarBKU99.pdf) is that there is little additional benefit to performing more than two choices. We would also creep closer to herding behavior with each additional selection we perform, so Envoy’s default configuration of the least request load balancer performs two choices.

https://github.com/envoyproxy/envoy/issues/11004 is a feature, not a bug with the LB algorithm.

Can you please update the description with what exactly the problem is with the selection probabilities and what effect this patch has on those selection probabilities? When it comes to modifying the behavior of the load balancing algorithms, there are many landmines we can step on, so the changes need more rigorous analysis.

**Comment by soulxu** [comment]
it seems this PR is waiting for @barroca response. thanks!

/wait

**Comment by tonya11en** [comment]
If the desired behavior is to unconditionally select the host with the least requests, it's fine to add a config parameter to use a full scan. This should not be the default (even for small host sets) or change the current behavior of any LBs.

**Comment by wbpcode** [comment]
@tonya11en I think we have a config option to this new feature. `enable_full_scan`. The full scaning doesn't not default behaviour.

**Comment by wbpcode** [comment]
@tonya11en I think we have a config option to this new feature. `enable_full_scan`. The full scaning doesn't not default behaviour.

**Comment by jkirschner-hashicorp** [comment]
Just to chime in with another use case for `enable_full_scan` that is mentioned in a [thread](https://github.com/envoyproxy/envoy/issues/4481) linked by @tonya11en :
> the fact that your backends are basically only capable of effectively handling a single request at full capacity and are CPU bound would lend itself to being more problematic. It seems totally reasonable to do full-scan as an alternative LB for least-loaded.

I'm trying to enable a scenario where each backend only accepts 1 long-lived (websocket) connection at a time, so we want Envoy to route connections to backends with 0 connections. The `enable_full_scan` mode submitted by @barroca could be used to enable that scenario (randomly selecting a backend with 0 connections since they would all have the "least requests").

This is all to say: I look forward to @barroca resubmitting the PR (combining the original and subsequent patches), and am very appreciative of all the discussion and review from maintainers!

**Comment by jkirschner-hashicorp** [comment]
@wbpcode, @tonya11en : By when would you need a resubmitted, combined PR for there to be a reasonable chance for this to land in Envoy 1.29.0, assuming review goes the way you expect? (I also understand if we've already passed that window.)

Thanks!

**Comment by jkirschner-hashicorp** [comment]
Just to document the intended contents of the combined PR:
1. https://github.com/envoyproxy/envoy/pull/29873: Original PR that introduced the “enable_full_scan” option.
2. https://github.com/envoyproxy/envoy/pull/30794: Don’t automatically use full scan mode even if ChoiceCount > number of hosts.
3. https://github.com/envoyproxy/envoy/pull/31146: Starts the scan at a random index in the host array to prevent hotspotting (in Zoom’s case: on the first host with 0 active connections).

My understanding is that 2 (#30794) was only a problem because the [first host was always selected](https://github.com/envoyproxy/envoy/pull/30768#discussion_r1387092315) if cluster stats are disabled. However, with the introduction of 3 (#31146), the host selected will be random if cluster stats are disabled.

It seems like the advantage of including just 1 and 3 without 2 is that you'll consider each host only once even if choice count > num hosts.

I defer to the maintainers on this point though (whether to include 1+3 or 1+2+3).

**Comment by wbpcode** [comment]
@jkirschner-hashicorp I think now you can re-submit the combined PR. And @tonya11en have implement a simulator in the #30818 to validate the problem and solution.

>  By when would you need a resubmitted, combined PR for there to be a reasonable chance for this to land in Envoy 1.29.0, assuming review goes the way you expect? (I also understand if we've already passed that window.)

Don't worry the time window. We can backport this PR even if it passed the window.

**Comment by tonya11en** [comment]
I have some bandwidth today to put up a PR that makes the selection method configurable (P2C vs. FULL_SCAN vs. ...). Let me know if you're already working on it and I'll just stand by to review when it's ready.

**Comment by barroca** [comment]
I haven't started anything else yet.

**Comment by jkirschner-hashicorp** [comment]
@barroca : In case you weren't in a position to move this forward at this time, I took a first pass yesterday at combining PRs 1 and 3 (omitting 2, because it seems unnecessary with the inclusion of 3): https://github.com/jkirschner-hashicorp/envoy/pull/1. I also made some small changes to the docs/comments.

Let me know how you'd like to proceed, happy to have you carry it forward as the original contributor!

**Comment by barroca** [comment]
Happy for take over and merge the changes with the combined PRs :) It is OSS after all and I've me my contributions already. I can focus on something else once I have time.

**Comment by tonya11en** [comment]
@jkirschner-hashicorp if you've started this I'll leave it up to you, then. The only thing I want to make sure we do is to configure the full scan with an enum representing the selection method (P2C vs. FULL_SCAN) instead of a boolean as found in the original PRs.

**Comment by jkirschner-hashicorp** [comment]
@tonya11en : I started, though initially with the expectation that I was just repackaging what was there. I've never worked with Envoy's source code before and am not familiar with some of the constraints (e.g., whether the `enable_full_scan` protobuf field needs to be kept, even though it was never in a released version, if we're switching to an Enum instead of a Bool). It may be more efficient for you to pick up from what's here rather than guide me in PR comments.

---

Either way, what are your thoughts on automatically using "full scan" if the number of choices configured is greater than the number of hosts? I was thinking of preserving that behavior, since the original motivation for stripping it out was that the first host was always selected if cluster stats were disabled (creating hotspots). That's no longer the case, now that the starting index of the full scan is random.

It's more efficient if the choice count configured is >= the number of hosts. And, if an external control plane is integrating with a version of `go-control-plane` that doesn't know about this new field yet, it could still take advantage of full scan mode by setting the choice count arbitrarily high.

**Comment by jkirschner-hashicorp** [comment]
I'll make a pass at converting the Bool to an enum. I now have a local build environment and got the least request load balancer tests passing.

**Comment by jkirschner-hashicorp** [comment]
Submitted a successor PR that uses an Enum rather than a Bool to specify the selection method (power of N choices or full scan): https://github.com/envoyproxy/envoy/pull/31507

File: source/common/upstream/load_balancer_impl.cc
Comment (tonya11en) [inline source/common/upstream/load_balancer_impl.cc:1307]: This is still going to be problematic as far as selection probabilities go. Consider some host vector with the following weights:

[9, 9, 1, 1, 1]

Choosing a random index to start the scan from would still choose the first host 80% of the time. The host at index 2 will only be picked 20% of the time, which seems unintuitive.

**Review by tonya11en — CHANGES_REQUESTED** [review]
This should probably be closed in favor of https://github.com/envoyproxy/envoy/pull/31507, which is based off of this and related patches.

**Comment by barroca** [comment]
closing in favour of [#31507](https://github.com/envoyproxy/envoy/pull/31507)

**Pull request closed without merging**
